api/client/crop_model.py

```diff
@@ -69,3 +119,83 @@ def compute_crop_weighted_series(self,
             # TODO: change metric to reflect it is weighted in this copy
             series_list.append(series)
         return pandas.concat(series_list)
+
+    def compute_gdd(self, tmin_series, tmax_series, base_temperature,
+                    start_date, end_date, min_temporal_coverage=0.5):
+        """Compute Growing Degree Days value from specific data series."""
+        self.add_single_data_series(tmin_series)
+        self.add_single_data_series(tmax_series)
+        df = self.get_df()
+        if df is None or df.empty:
+            raise Exception("Insufficient data for GDD in region {}".format(
+                region_name))
+        # For each day we want (t_min + t_max)/2, or more generally,
+        # the average temperature for that day.
+        tmean = df.loc[(df.item_id == tmax_series['item_id']) | \
+                       (df.item_id == tmin_series['item_id'])].groupby(
+                           ['region_id', 'metric_id', 'frequency_id',
+                            'start_date', 'end_date']).mean()
+        duration = datetime.strptime(end_date, '%Y-%m-%d') - \
+                   datetime.strptime(start_date, '%Y-%m-%d')
+        coverage_threshold = min_temporal_coverage * duration.days
+        if tmean.value.size < coverage_threshold:
+            raise Exception("Insufficient temporal coverage for GDD, " + \
+                            "{} < {} data points available".format(
+                                tmean.value.size, coverage_threshold))
+        gdd_values = tmean.value - base_temperature
```

Comment:
A few comments to ensure that the function works according to context:

The function currently allows for GDD to be unbounded; however, according to the source below, GDD shouldn't be negative, and should have an upper bound. At the very least I think we should not let it be negative, by modifying 

`gdd_values = tmean.value - base_temperature`

to 

`gdd_values = tmean.value.apply(lambda x: min(x - base_temperature, base_temperature))`

and also require base_temperature to be non-negative (or within a certain range), otherwise raise exception.

When there is no temperature data for a particular day, it would be better to raise a warning so that the user knows that the GDD is not accumulating due to missing data, instead of ineligible temperature range.

(Additionally, we could also require start and end date to be within the same growing season.)

Attached is my script where I tested a few input combinations. I modified the import path/generator syntax (`next(...)` instead of `(...).next()` )to make it work with Python 3.


Sources:
http://agron-www.agron.iastate.edu/courses/Agron541/classes/541/lesson02b/2b.1.1.html

Reply:
If the temperature at a given day, say `tmean.value`, is lower than the `base_temperature`, the contribution to GDD on that day should be zero.
Here's how I calculated GDD: ((
                                       (temp % up_thresh) * (
                                       temp < up_thresh
                                   ) + (temp > up_thresh) * up_thresh
                                   ) - lower_thresh) * (temp >= lower_thresh)

Reply:
Modified to match your expression.

Reply:
thanks!

- changed each gdd to >= 0 
- added support for upper_temperature_cap
- changed default min_temporal_coverage to 1.0 so by default we do not allow any missing days at all
- added warning when time range > 1 year